### ISSUE TITLE
Local (mac/windows/linux) testing should include alpn-boot

### ISSUE THREAD

**Issue by lesv** [GoogleCloudPlatform/appengine-java-vm-runtime#137]
Either always or based on env flag in app.yaml / appengine-web.xml:
      `<env-var name="ALPN_ENABLE" value="true" />`

I'm trying to do:
````
      <plugin>
        <groupId>com.google.appengine</groupId>
        <artifactId>gcloud-maven-plugin</artifactId>
        <version>${gcloud.plugin.version}</version>
        <configuration>
          <runtime>java</runtime>
          <jvm_flag>
            <param>-Xbootclasspath/p:${settings.localRepository}/org/mortbay/jetty/alpn/alpn-boot/${alpn.version}/alpn-boot-${alpn.version}.jar</param>
          </jvm_flag>
        </configuration>
      </plugin>
````
But I see that the code is running start.jar, so my `-Xbootclasspath` appears to be ignored.

[INFO] 2015-12-16 23:50:28.362:INFO:oejs.Server:main: jetty-9.2.10.v20150310

[INFO] Exception in thread "BigtableSession-startup-0" java.lang.NoClassDefFoundError: org/eclipse/jetty/alpn/ALPN$Provider
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkAlpnApplicationProtocolNegotiator$1.<init>(JdkAlpnApplicationProtocolNegotiator.java:26)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkAlpnApplicationProtocolNegotiator.<clinit>(JdkAlpnApplicationProtocolNegotiator.java:24)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkSslContext.toNegotiator(JdkSslContext.java:242)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkSslClientContext.<init>(JdkSslClientContext.java:263)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.SslContext.newClientContextInternal(SslContext.java:736)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.SslContextBuilder.build(SslContextBuilder.java:310)
[INFO] 	at com.google.cloud.bigtable.grpc.BigtableSession.createSslContext(BigtableSession.java:113)

I'm also trying to get [gRPC speedups](https://github.com/grpc/grpc-java/blob/master/SECURITY.md) using [netty-tcnative](http://netty.io/wiki/forked-tomcat-native.html) [(more)](http://netty.io/wiki/requirements-for-4.x.html) / OpenSSL to work on my Mac in this same session. Based on this [doc](https://cloud.google.com/bigtable/docs/using-maven)

**Comment by lesv** [comment]
What I see is:
````
[INFO] Exception in thread "BigtableSession-startup-0" java.lang.NoClassDefFoundError: org/eclipse/jetty/alpn/ALPN$Provider
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkAlpnApplicationProtocolNegotiator$1.<init>(JdkAlpnApplicationProtocolNegotiator.java:26)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkAlpnApplicationProtocolNegotiator.<clinit>(JdkAlpnApplicationProtocolNegotiator.java:24)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkSslContext.toNegotiator(JdkSslContext.java:242)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.JdkSslClientContext.<init>(JdkSslClientContext.java:263)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.SslContext.newClientContextInternal(SslContext.java:736)
[INFO] 	at com.google.bigtable.repackaged.io.netty.handler.ssl.SslContextBuilder.build(SslContextBuilder.java:310)
[INFO] 	at com.google.cloud.bigtable.grpc.BigtableSession.createSslContext(BigtableSession.java:113)
````

**Comment by lesv** [comment]
The startup line is:

`/Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk/Contents/Home/bin/java -Dgcloud.java.application=/Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT -Djetty.home=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime -Djetty.base=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk -Xbootclasspath/p:/Users/lesv/.m2/repository/org/mortbay/jetty/alpn/alpn-boot/8.1.3.v20150130/alpn-boot-8.1.3.v20150130.jar -jar /Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime/start.jar jetty.port=59345`

**Comment by ludoch** [comment]
@gregw: Maybe the way we use the start.jar is not compatible with this Xbootclasspath setup?

Need to move this in a property file?

**Comment by gregw** [comment]
@lesv is this a problem running locally or within a GKE image?   The entry point for the gke-debian-openjdk image is [docker-entrypoint.bash](https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/blob/master/gke-debian-openjdk/src/main/docker/docker-entrypoint.bash) , which sources [gke-env.bash](https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/blob/master/gke-debian-openjdk/src/main/docker/gke-env.bash), which if $ALPN_ENABLE is set runs  [alpn/format-env-appengine-vm.sh](https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/blob/master/gke-debian-openjdk/src/main/docker/alpn/format-env-appengine-vm.sh), which checks $ALPN_DISABLE and if null echoes the JVM bootclasspath args needed for ALPN.

So ALPN_ENABLE needs to be sent and ALPN_DISABLE needs to not be set.

This applies also to the gke-jetty image, except that it uses a modified [docker-entrypoint.bash](https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/blob/master/gke-jetty/src/main/docker/docker-entrypoint.bash)

**Comment by gregw** [comment]
@lesv @ludoch  so do you have the ALPN API class on the normal classpath?   You have to have the boot jar on the JVM bootpath, plus the API jar on the normal classpath

**Comment by lesv** [comment]
This is all running locally on my Mac.  Things have generally worked well using java-compat

**Comment by lesv** [comment]
It's not on the classpath.  In my pom I have the following:
```
    <dependency>
      <groupId>org.mortbay.jetty.alpn</groupId>
      <artifactId>alpn-boot</artifactId>
      <version>${alpn.version}</version>
      <scope>provided</scope>
    </dependency>
````

**Comment by gregw** [comment]
So here is the start.jar dry run line from running jetty with http2:
```
/usr/local/jvm/jdk1.8.0_51/jre/bin/java -Xbootclasspath/p:lib/alpn/alpn-boot-8.1.4.v20150727.jar -Djava.io.tmpdir=/tmp -Djetty.home=/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution -Djetty.base=/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/test -cp /home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/servlet-api-3.1.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-schemas-3.1.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-http-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-server-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-xml-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-util-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-io-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-alpn-server-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/http2/http2-common-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/http2/http2-hpack-9.3.7-SNAPSHOT.jar:/home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/http2/http2-server-9.3.7-SNAPSHOT.jar org.eclipse.jetty.xml.XmlConfiguration java.version=1.8.0_51 java.version.major=1 java.version.minor=8 java.version.revision=0 java.version.update=51 /home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/etc/jetty.xml /home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/etc/jetty-ssl.xml /home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/etc/jetty-ssl-context.xml /home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/etc/jetty-alpn.xml /home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/etc/jetty-http2.xml
```

Note that it has both: -Xbootclasspath/p:lib/alpn/alpn-boot-8.1.4.v20150727.jar and /home/gregw/src/jetty-9.3.x/jetty-distribution/target/distribution/lib/jetty-alpn-server-9.3.7-SNAPSHOT.jar

**Comment by gregw** [comment]
@lesv so that's the boot part, you also need the API part. Let me find the maven artifact to use....

**Comment by gregw** [comment]
@lesv cancel that - I'm wrong the API is also in the boot jar... re-evaluating....

**Comment by gregw** [comment]
@lesv @ludoch  I see the problem.  You are setting the boothpath on the JVM that runs start.jar, which is then execing a new JVM which does not have the bootpath on it.  You need to tell start.jar to include the boot path on what it runs.... just working that out....

**Comment by lesv** [comment]
Yep - suspected that -- Ideal would be that you use the full jar list that use use for Debian.  --dry-run output.

**Comment by gregw** [comment]
So instead of running
```
java -Xbootstuff -jar start.jar
```
you need to run
```
java -jar start.jar -Xbootstuff
```

Better yet, just do a 
```
java -jar ../start.jar --add-to-startd=alpn
```
and then start.jar will know to add the required boot path

**Comment by lesv** [comment]
Happy to do the add-to-startd, will it get the path to the jar?  Is there a way I can pass that?

**Comment by gregw** [comment]
@lesv if you use add-to-startd it will fetch the appropriate boot jar from maven, store it locally and use that... it might fetch it from your .m2 repo... but I'm not certain of that.

**Comment by lesv** [comment]
Trying it now.

**Comment by gregw** [comment]
@lesv if you want to depend on a specific version of alpn, you can do:
```
java -jar ../start.jar --add-to-startd=alpn-impl/alpn-1.8.0_60
```
look in $JETTY_HOME/modules/alpn-impl for the supported versions

**Comment by lesv** [comment]
It doesn't have any in the google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime/modules/...

I can pass a jar.

**Comment by lesv** [comment]
@gregw Can I set jetty.home to the jetty.war files in MVN that mvn jetty:run would use?  (I'm thinking of how to make this workable.

**Comment by gregw** [comment]
@lesv my examples have only been for the alpn jars. If you want other karts then you can either pass them on the command line to start. Jar out create your own module with those jars in it that can then be added. 

For mvn jetty run,  that is different yet again. I'll have to get back to you on how that works. Standby as I'm currently away from desk.

**Comment by lesv** [comment]
[INFO] Running python -S /Users/lesv/google-cloud-sdk/platform/google_appengine/dev_appserver.py --skip_sdk_update_check=true -A app /Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT/app.yaml --custom_entrypoint=**java
            -Dgcloud.java.application=/Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT 
            -Djetty.home=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime 
            -Djetty.base=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk 
            -jar /Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime/start.jar 
            --add-to-startd=/Users/lesv/.m2/repository/org/mortbay/jetty/alpn/alpn-boot/8.1.3.v20150130/alpn-boot-8.1.3.v20150130.jar
            jetty.port={port}**
[INFO] INFO     2015-12-17 02:12:25,361 application_configuration.py:431] No version specified. Generated version id: 20151217t021225
[INFO] INFO     2015-12-17 02:12:25,361 devappserver2.py:769] Skipping SDK update check.
[INFO] INFO     2015-12-17 02:12:25,416 api_server.py:205] Starting API server at: http://localhost:64571
[INFO] INFO     2015-12-17 02:12:25,425 dispatcher.py:197] Starting module "default" running at: http://localhost:8080
[INFO] INFO     2015-12-17 02:12:25,433 admin_server.py:116] Starting admin server at: http://localhost:8000
**[INFO] WARNING: ERROR: No known module for /Users/lesv/.m2/repository/org/mortbay/jetty/alpn/alpn-boot/8.1.3.v20150130/alpn-boot-8.1.3.v20150130.jar**

**Comment by gregw** [comment]
@lesv you can't use add-to-startd with a jar , also it is a command you run to configure the jetty base directory, so it takes effect on subsequent runs.

So you can either just do:
```
java -Djetty.base=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk -jar ../start.jar --add-to-startd=alpn
```
and then run:
```
python -S /Users/lesv/google-cloud-sdk/platform/google_appengine/dev_appserver.py --skip_sdk_update_check=true -A app /Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT/app.yaml --custom_entrypoint=java \
-Dgcloud.java.application=/Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT 
-Djetty.home=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime 
-Djetty.base=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk \
-jar /Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime/start.jar 
```

OR don't use the add mechanism and instead run:
```
python -S /Users/lesv/google-cloud-sdk/platform/google_appengine/dev_appserver.py --skip_sdk_update_check=true -A app /Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT/app.yaml --custom_entrypoint=java \
-Dgcloud.java.application=/Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT 
-Djetty.home=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime 
-Djetty.base=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk \
-jar /Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime/start.jar \
-Xbootclasspath/p:/Users/lesv/.m2/repository/org/mortbay/jetty/alpn/alpn-boot/8.1.3.v20150130/alpn-boot-8.1.3.v20150130.jar
```

**Comment by gregw** [comment]
@lesv or we can make a module for google apis, which can be based on the [gae module](gae module) in the compat image, which relies on all the gae jars being in lib/gae/*.jar

**Comment by janbartel** [comment]
@lesv, @gregw,

To use http2 with the maven jetty plugin, you need to do a bit of setup in the pom.xml, use a couple of xml files and also make sure that you're using the mvn jetty:run-forked goal to ensure that a new jetty process is forked with the alpn bootjar on it.

Steps are:

*  Follow the documentation on how to setup extra connectors for the jetty maven plugin documented here: http://www.eclipse.org/jetty/documentation/current/jetty-maven-plugin.html#configuring-jetty-container

* Also grab the following xml files from the distro: jetty-alpn.xml, jetty-http2.xml

* Ensure the plugin will apply these xml files in the correct order by adding this line to the plugin configuration:
```        
<jettyXml>jetty.xml,jetty-ssl.xml,jetty-ssl-context.xml,jetty-alpn.xml,jetty-http2.xml,jetty-https.xml</jettyXml>
```

* Tell the plugin where the appropriate alpn boot jar is, eg:

``` 
    <jvmArgs>-Xbootclasspath/p:alpn-boot-8.1.6.v20151105.jar</jvmArgs>
```

* Make sure you have the http2 and alpn jars added to the plugin dependencies:

``` 
     <dependencies>
           <dependency>
                 <groupId>org.eclipse.jetty</groupId>
                 <artifactId>jetty-alpn-server</artifactId>
                 <version>${jetty.version}</version>
          </dependency>
          <dependency>
                 <groupId>org.eclipse.jetty.http2</groupId>
                 <artifactId>http2-common</artifactId>
                 <version>${jetty.version}</version>
          </dependency>
          <dependency>
                 <groupId>org.eclipse.jetty.http2</groupId>
                 <artifactId>http2-hpack</artifactId>
                 <version>${jetty.version}</version>
          </dependency>
          <dependency>
                 <groupId>org.eclipse.jetty.http2</groupId>
                 <artifactId>http2-server</artifactId>
                 <version>${jetty.version}</version>
          </dependency>
        </dependencies>
```

You should now be able to do:  mvn jetty:run-forked and use alpn.

Hope that helps
Jan

**Comment by lesv** [comment]
I'll try this in the AM -- Thanks.

**Comment by lesv** [comment]
@gregw I think I tried the last one before, but I'll try again.  It seems like the -Xbootclasspath is being ignored due to something that is happening w/ start.java

If this doesn't work, I'll use the code that ludo had for running w/o start.java (passing the full list of jars)

**Comment by lesv** [comment]
BTW - just using --add-to-startd=alpn didn't work as the module isn't included in the SDK.

**Comment by lesv** [comment]
@gregw Ok, so I tried: 
````
[INFO] Running python -S /Users/lesv/google-cloud-sdk/platform/google_appengine/dev_appserver.py --skip_sdk_update_check=true -A app /Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT/app.yaml --custom_entrypoint=java -Dgcloud.java.application=/Users/lesv/github/cloud-bigtable-examples/java/managed-vm-gae/target/bigtable-hello-1.0-SNAPSHOT -Djetty.home=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime -Djetty.base=/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk -jar /Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/java-managed-vm/appengine-java-vmruntime/start.jar -Xbootclasspath/p:/Users/lesv/.m2/repository/org/mortbay/jetty/alpn/alpn-boot/8.1.3.v20150130/alpn-boot-8.1.3.v20150130.jar jetty.port={port}
[INFO] INFO     2015-12-17 19:40:49,121 application_configuration.py:431] No version specified. Generated version id: 20151217t194049
[INFO] INFO     2015-12-17 19:40:49,122 devappserver2.py:769] Skipping SDK update check.
[INFO] INFO     2015-12-17 19:40:49,187 api_server.py:205] Starting API server at: http://localhost:62507
[INFO] INFO     2015-12-17 19:40:49,193 dispatcher.py:197] Starting module "default" running at: http://localhost:8080
[INFO] INFO     2015-12-17 19:40:49,198 admin_server.py:116] Starting admin server at: http://localhost:8000
[INFO] WARNING: System properties and/or JVM args set.  Consider using --dry-run or --exec
[INFO] 2015-12-17 19:40:49.587:INFO::main: Logging initialized @372ms
[INFO] 2015-12-17 19:40:49.807:INFO:oejs.Server:main: jetty-9.2.10.v20150310
[INFO] 2015-12-17 19:40:49.821:WARN:oejsh.RequestLogHandler:main: !RequestLog
[INFO] 2015-12-17 19:40:49.824:INFO:oejdp.ScanningAppProvider:main: Deployment monitor [file:/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk/contexts/] at interval 1
[INFO] Dec 17, 2015 7:40:50 PM com.google.apphosting.vmruntime.VmMetadataCache getMetadata
[INFO] INFO: Meta-data 'attributes/gae_affinity' path retrieval error: metadata
[INFO] Dec 17, 2015 7:40:50 PM com.google.apphosting.vmruntime.VmMetadataCache getMetadata
[INFO] INFO: Meta-data 'attributes/gae_appengine_hostname' path retrieval error: metadata
[INFO] Dec 17, 2015 7:40:50 PM com.google.apphosting.vmruntime.VmMetadataCache getMetadata
[INFO] INFO: Meta-data 'attributes/gae_use_nginx_proxy' path retrieval error: metadata
[INFO] Dec 17, 2015 7:40:50 PM com.google.apphosting.vmruntime.VmMetadataCache getMetadata
[INFO] INFO: Meta-data 'attributes/gae_tmp_force_reuse_api_connection' path retrieval error: metadata
[INFO] 2015-12-17 19:40:50.230:WARN:oejx.XmlConfiguration:main: Config error at <Call name="init"><Arg><SystemProperty name="gcloud.java.application" default="."/></Arg><Arg><SystemProperty name="appengine_web_xml" default="WEB-INF/appengine-web.xml"/></Arg></Call> java.lang.reflect.InvocationTargetException in file:/Users/lesv/google-cloud-sdk/platform/google_appengine/google/appengine/tools/java/lib/jetty-base-sdk/contexts/root.xml
[INFO] 2015-12-17 19:40:50.230:WARN:oejd.DeploymentManager:main: Unable to reach node goal: started
````

I think I'm going to go back and do the equivalent of a --dry-run as I your giving me great suggestions for pure jetty, but what we ship in the sdk has mods, so I'm going to grab stuff from the appengine-java-vm-runtime repo as I've done in the past for Docker builds, then do a custom runtime against them.

**Comment by gregw** [comment]
@lesv that sounds like a reasonable approach.   So the fundamental advice is to ensure that the boot path is being set on the final JVM and not on the one that runs start.jar - using the --dry-run approach allows you to be absolutely clear and explicit on that front.

**Comment by gregw** [comment]
Can this issue be closed?

**Comment by lesv** [comment]
@ludoch This should be a feature for the SDK.  Not the whole thing I'm trying to do - using 9.2.xx and J8, but when we move to 9.3.x as part of jetty9-compat

**Comment by gregw** [comment]
closed for #280
